fix(commonjs)!: Correctly infer module name for any separator

pastelmind opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

<!-- the plugin(s) this PR is for -->

## Rollup Plugin Name: `commonjs`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [x] yes (_breaking changes will not be merged unless absolutely necessary_)
- [ ] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers: #923

### Description

When a module whose file name is `index.js` is resolved to a module ID containing a directory separator that is different from [`path.sep`](https://nodejs.org/docs/latest/api/path.html#path_path_sep) of the current platform, the commonjs plugin incorrectly infers the base name of the module. This causes the absolute path of the module to be embedded into the bundle, and prevents generating reproducible builds across different machines.

For example, with a module ID of `C:/Users/Phil/Documents/GitHub/test-vite-app/node_modules/react/index.js` on Windows, the plugin generates code like this:
```js
var C__Users_Phil_Documents_GitHub_testViteApp_node_modules_react = { exports: {} };
var react_production_min = {};
```

This PR fixes the problem by using `path.basename()` to extract the final path component instead of splitting the string with `path.sep`.

With the fix, the plugin generates code like this:
```js
var react = { exports: {} };
var react_production_min = {};
```

Closes #923

Chucky-ux reviewed (commented):
I think we are on the right way. 😊

pastelmind commented:
I added a test. Unfortunately, it works only on Windows, where replacing backslashes with slashes is perfectly fine. On Linux, the backslash is treated as a regular character, which means that we'd get an error if we replaced slashes with backslashes.

pastelmind commented:
@danielgindi I apologize for pinging you, but this PR hasn't even got a chance to run workflows yet. Could you please take a look?

pastelmind commented:
@danielgindi Thanks for letting the CI workflow happen, but it failed and I'm not sure why. The log is at https://github.com/rollup/plugins/runs/3150391648, here's an excerpt:
> Run pnpm run ci:test --filter ...[36a461e2f9775339554b7c76b1a2a8db408b9635]
>  pnpm run ci:test --filter ...[36a461e2f9775339554b7c76b1a2a8db408b9635]
>  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
> ERROR  Filtering by changed packages failed. fatal: bad object 36a461e2f9775339554b7c76b1a2a8db408b9635
> Error: Process completed with exit code 1.

I rebased the commit onto the current `master` and force-pushed, but now the workflow needs approval again. I'm not familiar with the CI process--could you please help me troubleshoot this?

shellscape commented:
@pastelmind looks like your tests all pass now.  I've requested review from a few folks to get eyes on this for you.

danielgindi commented:
To me this looks much more resilient than the current code. Using core path functions handles this stuff for us.

shellscape commented:
![thanks!](https://i.imgur.com/BrUW1rG.jpg)
